### PR TITLE
Fix creating a lot of ephemeral ports when stopping GlassFish. Fix restart on fast machines

### LINKED ISSUE #25292: Ephemeral ports are exhausted if stopping the DAS takes a long time
## Environment Details
* GlassFish Version (and build number): Eclipse GlassFish 7.0.20
* JDK version: OpenJDK 21.0.4
* OS: Windows 10
* Database: none

----------

## Problem Description
If it takes a long time to stop the DAS, many ephemeral ports are opened.
In the worst case, the ephemeral ports provided by the operating system are exhausted.

## Steps to reproduce

1. Compile the following lifecycle listener including a long-running task.
```java
import com.sun.appserv.server.LifecycleEvent;
import com.sun.appserv.server.LifecycleListener;
import com.sun.appserv.server.ServerLifecycleException;

public class TestLifecycleListener implements LifecycleListener {
    @Override
    public void handleEvent(LifecycleEvent event) throws ServerLifecycleException {
        if (event.getEventType() == LifecycleEvent.SHUTDOWN_EVENT) {
            try {
                // do long-running task
                Thread.sleep(15000);
            } catch (InterruptedException e) {
            }
        }
    }
}
```
```
> javac -cp C:\glassfish7\glassfish\modules\glassfish-api.jar TestLifecycleListener.java
```

2. Enable the lifecycle listener.
```
> asadmin start-domain
> asadmin create-lifecycle-module --classname TestLifecycleListener --classpath C:/work/lifecycle TestLifecycleListener
> asadmin stop-domain
> asadmin start-domain
```

3. Stop the DAS.
```
> asadmin stop-domain
Waiting for the domain to stop ...........
Waiting finished after 15,668 ms.
Command stop-domain executed successfully.
```

4. Just after stoppiong the DAS, by checking `netstat`, we can see that many ephemeral ports are consumed.

```
> netstat -ano
-- snip --
  TCP         127.0.0.1:49152        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49153        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49154        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49155        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49156        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49157        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49158        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49159        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49160        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:49161        127.0.0.1:4848         TIME_WAIT       0
-- snip --
  TCP         127.0.0.1:65526        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65527        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65528        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65529        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65530        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65531        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65532        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65533        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65534        127.0.0.1:4848         TIME_WAIT       0
  TCP         127.0.0.1:65535        127.0.0.1:4848         TIME_WAIT       0
```

## Impact of Issue
If the ephemeral ports are exhausted, new TCP connections requiring ephemeral ports cannot be made.

### LINKED ISSUE #25295: The asadmin restart-domain may cause race conditions
Reproducer - run our own tests on a powerful cpu, in my case AMD Ryzen 9 7945HX.

```
mvn clean install -pl :admin-tests
```
Used OS, Java:
```
mvn -version
Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937)
Maven home: /home/dmatej/work/apache-maven-3.9.9
Java version: 21.0.5, vendor: Eclipse Adoptium, runtime: /usr/lib/jvm/jdk-21.0.5+11
Default locale: cs_CZ, platform encoding: UTF-8
OS name: "linux", version: "6.11.0-108013-tuxedo", arch: "amd64", family: "unix"
```

Failed tests - well reproducible, however individual executions passed:
```
[ERROR]   LoggingRestITest.fillUpLog:59 
Expected: asadmin succeeded
     but: was <Waiting finished after 12 ms.

Process is still running, timeout 30000 ms exceeded.>
[ERROR] Errors: 
[ERROR]   DomainITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   ExternalResourceITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   JmsITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   ProvidersITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   ResourceRefITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   ServerPropertyITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   StatusGeneratorITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[ERROR]   ZeroConfigITest>RestTestBase.initialize:81 » Processing java.net.ConnectException: Connection refused
[INFO] 
[ERROR] Tests run: 122, Failures: 7, Errors: 8, Skipped: 0
```

server.log
```
[2024-12-22T18:58:57.565673+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00172] [jakarta.enterprise.web] [tid: _ThreadID=130 _ThreadName=admin-listener(1)] [levelValue: 800] [[
  Loading application [HttpServerNameITest] at [/HttpServerNameITest]]]

[2024-12-22T18:58:57.620385+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [] [jakarta.enterprise.system.core] [tid: _ThreadID=130 _ThreadName=admin-listener(1)] [levelValue: 800] [[
  HttpServerNameITest was successfully deployed in 104 milliseconds.]]

[2024-12-22T18:58:57.895114+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=191 _ThreadName=pool-17-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:57.897331+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=191 _ThreadName=pool-17-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:58.113013+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=195 _ThreadName=pool-20-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:58.114910+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=195 _ThreadName=pool-20-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:58.308350+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:58.310636+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:58.506665+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=199 _ThreadName=pool-24-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:58.508402+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=199 _ThreadName=pool-24-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:58.698124+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:58.700068+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 1 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:58.905316+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=191 _ThreadName=pool-17-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:58.907582+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=191 _ThreadName=pool-17-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:59.105948+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=199 _ThreadName=pool-24-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:59.108107+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=199 _ThreadName=pool-24-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 1 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:59.314237+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:59.316298+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 1 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:59.528754+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=193 _ThreadName=pool-18-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:59.531023+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=193 _ThreadName=pool-18-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 1 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:59.722815+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:59.724506+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=194 _ThreadName=pool-19-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:58:59.915946+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:58:59.917832+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:59:00.120409+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:59:00.122837+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=196 _ThreadName=pool-21-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:59:00.333513+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=195 _ThreadName=pool-20-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:59:00.335582+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=195 _ThreadName=pool-20-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:59:00.543317+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [AS-WEB-GLUE-00198] [jakarta.enterprise.web] [tid: _ThreadID=193 _ThreadName=pool-18-thread-1] [levelValue: 800] [[
  Created HTTP listener http-listener-1 on host/port 0.0.0.0:8080]]

[2024-12-22T18:59:00.545720+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [NCLS-CORE-00087] [jakarta.enterprise.system.core] [tid: _ThreadID=193 _ThreadName=pool-18-thread-1] [levelValue: 800] [[
  Grizzly Framework 4.0.2 started in: 0 ms - bound to [/0.0.0.0:8080]]]

[2024-12-22T18:59:00.769964+01:00] [GF 7.0.21-SNAPSHOT] [INFO] [] [org.glassfish.web.loader.WebappClassLoader] [tid: _ThreadID=130 _ThreadName=admin-listener(1)] [levelValue: 800] [[
  close(), this:
org.glassfish.web.loader.WebappClassLoader@19778c77[name=/HttpServerNameITest], urls=[
file:/home/dmatej/work/repo/git/glassfish-new/appserver/tests/admin/tests/target/glassfish7/glassfish/domains/domain1/applications/HttpServerNameITest/WEB-INF/classes
file:/home/dmatej/work/repo/git/glassfish-new/appserver/tests/admin/tests/target/glassfish7/glassfish/domains/domain1/generated/ejb/HttpServerNameITest/
file:/home/dmatej/work/repo/git/glassfish-new/appserver/tests/admin/tests/target/glassfish7/glassfish/domains/domain1/generated/HttpServerNameITest/
][delegate=true, context=/HttpServerNameITest, status=RUNNING, antiJARLocking=false, securityManager=false, packageDefinitionSecurityEnabled=false, repositories=RepositoryManager[WEB-INF/classes/], notFound.size=2, pathTimestamps.size=0, resourceEntryCache.size=0]]]

```

### PR DESCRIPTION
**Pull request by dmatej**
Fixes #25295 and #25292
Replaces #25293 

There were several issues, see individual commits. The main problem was that the startup can be faster than shutdown and then they could collide on ports and files. The most problematic was the debug port which is enabled since the JVM startup until the very end.

On my new machine it was reproducible in some 80% of executions.

Solution for #25292:
* Instead of busy spinning on remote port we now open the connection and wait until connection is disconnected.

For #25295 was needed also
* Move startup to shutdown hooks
* The startup hook waits for the end of other glassfish shutdown hooks (detected by name)
* Logging is explicitly stopped
* For extreme cases I added additional logging for dying and starting process which can be enabled by setting an environment option `export AS_RESTART_LOGFILES=true`;

Note
* See https://bugs.openjdk.org/browse/JDK-8284282 - applies to our Jenkins and most docker containers too. Terminated GlassFish instances become zombies, then `handle#isAlive` returns true and `onExit.get` is still blocked.

### CODE REVIEW

**Comment by dmatej** [comment]
Heuréka! And now I see why we had that weird code checking of `info()` ... Jenkins/k8s/docker doesn't reap zombies and then `onExit().get()` blocks forever.
https://bugs.openjdk.org/browse/JDK-8284282?jql=status%20in%20(Closed%2C%20Submitted)%20AND%20text%20~%20"ProcessHandle"

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+
+    private static ProcessBuilder prepareStartup(Instant startTime, String classpath, String[] sysprops, String classname,
+        String[] args) {
+        final Path javaExecutable = detectJavaExecutable();
+        final List<String> cmdline = new ArrayList<>();
+        if (!OS.isWindows()) {
+            cmdline.add("nohup");
+        }
+        cmdline.add(javaExecutable.toFile().getAbsolutePath());
```
Comment (avpinchuk) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java:198]: Why not use just `Path.toAbsolutePath()` or `Path.toAbsolutePath().normalize()`?

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        this.logFile = getLogFileOld(startTime);
+        this.builder = prepareStartup(startTime, classpath, sysProps, classname, args);
+    }
+
+
+    @Override
+    public void run() {
+        try {
+            // We do want to be the last thread alive.
+            Thread.onSpinWait();
```
Comment (avpinchuk) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java:84]: Is spin waiting neccessary here?

**Review by OndroMih — APPROVED** [review]

**Comment by OndroMih** [comment]
I renamed this PR so that it reflects all things that were fixed. The title will appear in the release notes so it's good if it clearly describes all that this PR adds/fixes.

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java (hunk 1 above)
Comment (dmatej) [reply nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java:198]: It doesn't have same result on all Windows if I remember well. Path always uses /, on current Windows "it depends".

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java (hunk 2 above)
Comment (dmatej) [reply nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java:84]: Yes, originally I implemented your suggestion with onExit, but it freezes on Jenkins - it heavily depends on the OS reaper but not all systems have them.
I added some documentation for that and link to jdk jira. I had yet few others to Stackoverflow, and spent two says with that.
Locally it worked well, but failed on jenkins. 

Maybe you will find something I missed, I don't know.

Comment (avpinchuk) [reply nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java:84]: Thank you for answers. I have no more questions. Great job!

**Review by avpinchuk — APPROVED** [review]

**Pull request merged**